### PR TITLE
[GenericPoseGraph] Add RelativePoseConstraint2D.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

File: cartographer/pose_graph/constraint/constraint.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace cartographer {
+namespace pose_graph {
+
+using ConstraintId = std::string;
+
+class Constraint {
+ public:
+  explicit Constraint(const ConstraintId& id) : constraint_id_(id) {}
+  ~Constraint() = default;
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/constraint.h]: virtual

File: cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "cartographer/common/make_unique.h"
+
+namespace cartographer {
+namespace pose_graph {
+namespace {
+
+template <typename MapType, typename KeyType = typename MapType::key_type,
+          typename ValueType = typename MapType::mapped_type>
+ValueType* FindOrNull(MapType& map, const KeyType& key) {
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc]: Not in this CL but can we have that somewhere in common?

File: cartographer/pose_graph/constraint/relative_pose_constraint_2d.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  void AddToOptimizer(Nodes* nodes, ceres::Problem* problem) const final;
+
+ protected:
+  proto::CostFunction ToCostFunctionProto() const final;
+
+ private:
+  NodeId first_;
+  NodeId second_;
+  std::unique_ptr<RelativePoseCost2D> ceres_cost_;
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/relative_pose_constraint_2d.h]: Why does this need to be a unique_ptr?

File: cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc (hunk 2 above)
Comment (pifon2a) [reply cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc]: Definitely.

File: cartographer/pose_graph/constraint/constraint.h (hunk 1 above)
Comment (pifon2a) [reply cartographer/pose_graph/constraint/constraint.h]: Ooops

File: cartographer/pose_graph/constraint/relative_pose_constraint_2d.h (hunk 3 above)
Comment (pifon2a) [reply cartographer/pose_graph/constraint/relative_pose_constraint_2d.h]: It does not :). Done.

File: cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc [hunk 4]
```diff
@@ -70,7 +70,7 @@ void RelativePoseConstraint2D::AddToOptimizer(Nodes* nodes,
   if (second_node->constant()) {
     problem->SetParameterBlockConstant(second_pose->data());
   }
-  problem->AddResidualBlock(ceres_cost_.get(), nullptr /* loss function */,
+  problem->AddResidualBlock(&ceres_cost_, nullptr /* loss function */,
```
Comment (pifon2a) [inline cartographer/pose_graph/constraint/relative_pose_constraint_2d.cc]: nah, does not work:

relative_pose_constraint_2d.cc:73:29: error: invalid conversion from ‘const ceres::CostFunction*’ to ‘ceres::CostFunction*’ [-fpermissive]
   problem->AddResidualBlock(&ceres_cost_, nullptr /* loss function */,

**Pull request merged**